crates/interpreter/src/interpreter/shared_memory.rs

```diff
@@ -0,0 +1,297 @@
+            None => debug_unreachable!("set_byte OOB: {index}; len: {}", self.len()),
+        }
+    }
+
+    /// Sets the given `value` to the memory region at the given `offset`.
+    ///
+    /// Panics on out of bounds.
+    #[inline(always)]
+    #[cfg_attr(debug_assertions, track_caller)]
+    pub fn set_u256(&mut self, offset: usize, value: U256) {
+        self.set(offset, &value.to_be_bytes::<32>());
+    }
+
+    /// Set memory region at given `offset`.
+    ///
+    /// Panics on out of bounds.
+    #[inline(always)]
+    #[cfg_attr(debug_assertions, track_caller)]
+    pub fn set(&mut self, offset: usize, value: &[u8]) {
+        if !value.is_empty() {
+            self.slice_mut(offset, value.len()).copy_from_slice(value);
+        }
+    }
+
+    /// Set memory from data. Our memory offset+len is expected to be correct but we
+    /// are doing bound checks on data/data_offeset/len and zeroing parts that is not copied.
+    ///
+    /// Panics on out of bounds.
+    #[inline(always)]
+    #[cfg_attr(debug_assertions, track_caller)]
+    pub fn set_data(&mut self, memory_offset: usize, data_offset: usize, len: usize, data: &[u8]) {
+        if data_offset >= data.len() {
+            // nullify all memory slots
+            self.slice_mut(memory_offset, len).fill(0);
+            return;
+        }
+        let data_end = min(data_offset + len, data.len());
+        let data_len = data_end - data_offset;
+        debug_assert!(data_offset < data.len() && data_end <= data.len());
+        let data = unsafe { data.get_unchecked(data_offset..data_end) };
+        self.slice_mut(memory_offset, data_len)
+            .copy_from_slice(data);
+
+        // nullify rest of memory slots
+        // Safety: Memory is assumed to be valid. And it is commented where that assumption is made
+        self.slice_mut(memory_offset + data_len, len - data_len)
+            .fill(0);
+    }
+
+    /// Copies elements from one part of the memory to another part of itself.
+    ///
+    /// Panics on out of bounds.
+    #[inline(always)]
+    #[cfg_attr(debug_assertions, track_caller)]
+    pub fn copy(&mut self, dst: usize, src: usize, len: usize) {
+        self.get_current_slice_mut()
+            .copy_within(src..src + len, dst);
+    }
+
+    /// Get a refernce to the memory of the current context
+    #[inline(always)]
+    fn get_current_slice(&self) -> &[u8] {
+        // Safety: if it is a valid pointer to a slice of `self.data`
+        unsafe { &*self.current_slice }
+    }
+
+    /// Get a mutable refernce to the memory of the current context
+    #[inline(always)]
+    fn get_current_slice_mut(&mut self) -> &mut [u8] {
+        // Safety: it is a valid pointer to a slice of `self.data`
+        unsafe { &mut *self.current_slice }
+    }
+
+    /// Calculates memory allocation upper bound using
+    /// https://2π.com/22/eth-max-mem
+    #[inline(always)]
+    fn calculate_upper_bound(gas_limit: u64) -> u64 {
```

Comment:
And move to the top of the impl block, this could probably be useful elsewhere or downstream

```suggestion
    #[inline]
    pub fn calculate_upper_bound(gas_limit: u64) -> u64 {
```